LocalProgram refactoring

xclaesse opened the pull request:
Reapply patchset from https://github.com/mesonbuild/meson/pull/15107.

xclaesse committed:
Add support for LocalProgram to custom_target() and run_target()

xclaesse committed:
Add support for LocalProgram to configure_file() and run_command()

xclaesse committed:
Add support for LocalProgram to test() and benchmark()

xclaesse committed:
Add support for LocalProgram to generator()

xclaesse committed:
Add support for LocalProgram to override_find_program()

xclaesse committed:
Replace OverrideExecutable and OverrideProgram with LocalProgram

xclaesse committed:
Add common ABC for ExternalProgram and LocalProgram

xclaesse committed:
Add support for LocalProgram to add_*_script()

xclaesse commented:
@jpakkane @eli-schwartz can you reviews this please?

eli-schwartz reviewed (changes requested):
As I said previously:

Adding a new toplevel function which is an existing function with very minor behavior differences does not seem warranted to me. The main justification for a new function appears to be simply that it won't searchh PATH, but this is redundant with passing a files() to `find_program()`. I think this is a bad reason to add a new toplevel function.

Design wise, `depends:` and `interpreter` could just as easily be added to find_program directly, and there are potential use cases for the former even for programs found via PATH.

But I also think that this is unnecessary to allow interpreter as a kwarg at all. Existing users always pass an array around to things like

```meson
my_gen = [python, find_program(files('my_gen.py'))]
custom_target(command: my_gen + [....])
```

We should work with that and allow array overrides directly to `meson.override_find_program()`. I believe that will be a lot more intuitive than having "split-brain" interfaces and new toplevel functions depending on whether the result is intended to be passed to override_find_program.

(As a bonus, it also means for example that you can do override_find_program in an `if meson.version().version_compare`, the same way override_dependency was "simple to support" in a backwards compatible manner. That doesn't necessarily mean that I think new design should be based on what's easier to write backports using -- I just think it's very convenient that the thing I believe is a better design anyway, is also easier to use with backports.)

Passing an array to override_find_program means we can easily support interpreter options, which your original design didn't allow at all. I see that this new version does allow it, but it looks forced in and you specifically have to document that it works despite the obvious assumption (and the kwarg name implies it doesn't work, which is awkward UX). You still don't support arguments after the program, which prevents some use cases that would "automatically work" when just allowing to export an array, and we wouldn't have to take an opinionated stance about whether users should be allowed to do it. I can think of reasons why it would be perfectly fine to do, but it seems like you want a top-level function with an `interpreter` kwarg in part just because you want to ban something that is a "niche power tool" that you are afraid people will use in ways you find inelegant, and they could already do except more difficult. I don't see the point of this limitation.

...

By the way I would still like to see this support CustomTarget first, which I think we all agree should be made to work, and which you said this refactoring could be in support of instead. That would allow us to merge the groundwork independently and make it easier to reason about the scope of this proposal.

xclaesse committed:
Add LocalProgram object

xclaesse committed:
Rename BaseProgram to Program and use it where possible

When both ExternalProgram and LocalProgram are accepted, use Program
instead.

xclaesse committed:
LocalProgram: fix full_path() and cmd_array()

xclaesse committed:
BaseProgram: Define for_machine

xclaesse commented:
> * creating the base class and then keeping the Union in most of the annotations is confusing

Yeah, I still had a commit for that locally that mypy wasn't happy about, but I found the reason now. Fixed.

> * `LocalProgram.get_command()` cannot work (and therefore `BaseProgramHolder.cmd_array_method()` is broken)

Fixed and made a unit test for it, because `prog.cmd_array()` with an exe override is broken on master too.

> * BaseProgram is missing the `for_machine` field

done
